substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java

```diff
@@ -0,0 +1,234 @@
+            return this;
+        }
+
+        CVTypeProcedureRecord argList(CVTypeArglistRecord leaf) {
+            this.argList = leaf;
+            return this;
+        }
+
+        @Override
+        public int computeSize(int initialPos) {
+            return computeContents(null, initialPos);
+        }
+
+        @Override
+        public int computeContents(byte[] buffer, int initialPos) {
+            int pos = CVUtil.putInt(returnType, buffer, initialPos);
+            pos = CVUtil.putByte((byte) 0, buffer, pos); /* callType */
+            pos = CVUtil.putByte((byte) 0, buffer, pos); /* funcAttr */
+            pos = CVUtil.putShort((short) argList.getSize(), buffer, pos);
+            pos = CVUtil.putInt(argList.getSequenceNumber(), buffer, pos);
+            return pos;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("LF_PROCEDURE 0x%04x ret=0x%04x arg=0x%04x ", getSequenceNumber(), returnType, argList.getSequenceNumber());
+        }
+
+        @Override
+        public int hashCode() {
+            int h = type;
+            h = 31 * h + returnType;
+            h = 31 * h + argList.hashCode();
+            /* callType and funcAttr are always zero so do not add them to the hash */
+            return h;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            CVTypeProcedureRecord other = (CVTypeProcedureRecord)obj;
+            return super.equals(obj) && this.returnType == other.returnType && this.argList == other.argList;
+        }
+    }
+
+    static final class CVTypeArglistRecord extends CVTypeRecord {
+
+        ArrayList<Integer> args = new ArrayList<>();
+
+        CVTypeArglistRecord() {
+            super(LF_ARGLIST);
+        }
+
+        CVTypeArglistRecord add(int argType) {
+            args.add(argType);
+            return this;
+        }
+
+        @Override
+        public int computeSize(int initialPos) {
+            return initialPos + Integer.BYTES + Integer.BYTES * args.size();
+        }
+
+        @Override
+        public int computeContents(byte[] buffer, int initialPos) {
+            int pos = CVUtil.putInt(args.size(), buffer, initialPos);
+            for (Integer at : args) {
+                pos = CVUtil.putInt(at, buffer, pos);
+            }
+            return pos;
+        }
+
+        int getSize() {
+            return args.size();
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder s = new StringBuilder(String.format("LF_ARGLIST 0x%04x [", getSequenceNumber()));
+            for (Integer at : args) {
+                s.append(String.format(" 0x%04x", at));
+            }
+            s.append("])");
+            return s.toString();
+        }
+
+        @Override
+        public int hashCode() {
+            int h = type;
+            h = h * 31 + args.size();
+            for (Integer r : args) {
+                h = 31 * h + r.hashCode();
+            }
```

Comment:
Why didn't you use `args.hashCode` here?

Reply:
Done.